File: testsuite/tests/backtrace/backtrace_c_exn.ml
(* TEST
   modules = "backtrace_c_exn_.c"
   flags = "-g"
   ocamlrunparam += ",b=1"
*)

(* https://github.com/ocaml-multicore/ocaml-multicore/issues/498 *)
external stubbed_raise : unit -> unit = "caml_498_raise"

let raise_exn () = failwith "exn"

let () = Callback.register "test_raise_exn" raise_exn

let () =
  try
    stubbed_raise ()
  with
  | exn ->
    Printexc.to_string exn |> print_endline;
    Printexc.print_backtrace stdout

File: testsuite/tests/backtrace/backtrace_c_exn_.c
#include <caml/callback.h>
#include <caml/memory.h>

void caml_498_raise(void) {
    CAMLparam0 ();
    const value *cl;

    cl = caml_named_value("test_raise_exn");

    if (cl != NULL)
      caml_callback(*cl, Val_unit);

    CAMLreturn0;
}
